builder/bundles.go

```diff
@@ -480,30 +515,149 @@ func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int)
+	rpm2Cmd.Env = os.Environ()
+
+	cpioCmd := exec.Command("cpio", "-id")
+	cpioCmd.Env = os.Environ()
+	cpioCmd.Dir = baseDir
+
+	pr, pw := io.Pipe()
+	rpm2Cmd.Stdout = pw
+	cpioCmd.Stdin = pr
+
 	var err error
-	baseDir := filepath.Join(buildVersionDir, "full")
-	rpmList := []string{}
 
-	if len(bundle.AllPackages) > 0 {
-		// There were packages directly included for this bundle so
-		// install to full chroot. This check is necessary so we don't
-		// call dnf install with no package listed.
-		for p := range bundle.AllPackages {
-			rpmList = append(rpmList, p)
+	err = rpm2Cmd.Start()
+	if err != nil {
+		return fmt.Errorf("rpm2cpio failed with %v", err)
+	}
+	err = cpioCmd.Start()
+	if err != nil {
+		return fmt.Errorf("cpio failed with %v", err)
+	}
+
+	err = rpm2Cmd.Wait()
+	if err != nil {
+		return fmt.Errorf("rpm2cpio failed with %v", err)
+	}
+
+	err = pw.Close()
+	if err != nil {
+		return fmt.Errorf("  writer pipe failed to close:  %v ", err)
+	}
+
+	err = cpioCmd.Wait()
+	if err != nil {
+		return fmt.Errorf("cpioCmd failed with %v", err)
+	}
+
+	err = pr.Close()
+	if err != nil {
+		return fmt.Errorf(" reader pipe failed to close: %v ", err)
+	}
+
+	return err
+}
+
+func installBundleToFull(packagerCmd []string, baseDir string, bundle *bundle, downloadRetries int, numWorkers int, localPath string) error {
+	var err error
+	var wg sync.WaitGroup
+	rpmCh := make(chan string)
+
+	var errCh error
+	var rpmList []string
+	var rpmPath string
+
+	if len(bundle.AllRpmPackages) < numWorkers {
+		numWorkers = len(bundle.AllRpmPackages)
+	}
+	errorCh := make(chan error, numWorkers)
+	wg.Add(numWorkers)
+
+	rpmWorker := func() {
+		var e error
+		for rpm := range rpmCh {
+			for i := 0; i < extractRetries; i++ {
+				e = extractRpm(baseDir, rpm)
+				if e == nil {
+					break
+				}
+			}
+			if e != nil {
+				errorCh <- e
+				break
+			}
 		}
+		wg.Done()
+	}
 
-		// Retry RPM downloads to avoid timeout failures due to slow network
-		_, err = downloadRpms(packagerCmd, rpmList, baseDir, downloadRetries)
-		if err != nil {
-			return err
+	for i := 0; i < numWorkers; i++ {
+		go rpmWorker()
+	}
+
+	if !Offline {
+		for rpm := range bundle.AllRpmPackages {
+			rpmFull := filepath.Join(localPath, rpm)
+			if _, err = os.Stat(rpmFull); os.IsNotExist(err) {
+				rpm1 := strings.TrimSuffix(rpm, ".rpm")
+				rpmList = append(rpmList, rpm1)
+			}
+		}
+		if len(rpmList) > 0 {
+			rpmPath, err = downloadRpm(packagerCmd, rpmList, baseDir, downloadRetries)
+			if err != nil {
+				return err
+			}
 		}
+	}
 
-		args := merge(packagerCmd, "--installroot="+baseDir, "install")
-		args = append(args, rpmList...)
-		_, err = helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
-		if err != nil {
-			return err
+	// feed the channel
+	for rpm := range bundle.AllRpmPackages {
+		if rpmMap[rpm] {
+			continue
+		}
+		rpmFull := filepath.Join(localPath, rpm)
+		if _, err = os.Stat(rpmFull); os.IsNotExist(err) {
```

Comment:
This block doesn't do anything. Plus, if you follow my suggestion above, this will never be true.